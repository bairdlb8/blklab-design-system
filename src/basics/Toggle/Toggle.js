import React from 'react'
import PropTypes from 'prop-types'

import './Toggle.scss'

const Toggle = ({ name = '', label = '', leftLabel = '', rightLabel = '', className = '' }) => {
  return (
    <fieldset aria-label={label}>
      <legend>{label}</legend>
      <div className="ds-toggle" role="radiogroup">
        <label htmlFor="left" className="sr-only">
          {leftLabel}
        </label>
        <span className="ds-toggle__wrapper">
          <input type="radio" name={name} id="left" aria-label={leftLabel} />
          <input type="radio" name={name} id="right" aria-label={rightLabel} />
          <span aria-hidden="true" className="ds-toggle__background"></span>
          <span aria-hidden="true" className="ds-toggle__switcher"></span>
        </span>
        <label htmlFor="right" className="sr-only">
          {rightLabel}
        </label>
      </div>
    </fieldset>
  )
}

Toggle.propTypes = {}

export default Toggle
