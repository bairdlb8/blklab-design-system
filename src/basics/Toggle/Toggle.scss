.ds-toggle {
  --baseSize: 12px;

  label {
    font-size: calc(var(--baseSize) * 1.75);
    margin: 0 0.5em 0;
    cursor: pointer;
  }

  input[type='radio'] {
    display: inline-block;
    margin-right: -2px;
    width: 50%;
    height: 100%;
    opacity: 0;
    position: relative;
    z-index: 1;
    cursor: pointer;

    &:focus ~ .ds-toggle__switcher {
      border: 3px solid hsla(333, 40%, 45%, 1);
      top: -2px;
    }
  }
}

.ds-toggle__wrapper {
  font-size: var(--baseSize);
  display: inline-block;
  vertical-align: middle;
  width: 7em;
  height: 3.5em;
  border-radius: 3.5em; /* = height */
  border: 1px solid #ddd;
  position: relative; // create positioning context for ::after
}
.ds-toggle__switcher {
  // the toggle
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 100%;
  width: 50%;
  height: 100%;
  border-radius: 50%;
  background-color: hsla(333, 60%, 60%, 1);
  transition: all 0.25s ease-out;
  z-index: 2;
}

.ds-toggle__background {
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 0;
  border-radius: 3.5em;
  background-color: #eee;
  transition: all 0.1s ease-out;
}

#right:checked ~ .ds-toggle__switcher {
  top: 0;
  right: 0;
  left: 50%;
}

#right:checked ~ .ds-toggle__background {
  background-color: #333;
}

a {
  display: block;
  margin: 2em 0;
  padding: 0.5em;
  box-sizing: border-box;

  &:focus {
    outline: 3px solid #000;
  }
}

/* support Windows High Contrast Mode. Credit: Adrian Roselli https://twitter.com/aardrian/status/1021372139990134785 */
@media screen and (-ms-high-contrast: active) {
  .ds-toggle__switcher {
    background-color: windowText;
  }
}

fieldset {
  border: none;
}
