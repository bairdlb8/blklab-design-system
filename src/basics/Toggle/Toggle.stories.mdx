import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks'

import Toggle from './Toggle'

<Meta title="Basics / Toggle" component={Toggle} />

# Toggle

## Add some text here

export const Template = (args) => {
  return <Toggle {...args} />
}

<Preview withToolbar>
  <Story name="Primary" args={{ label: 'Test', name: 'test', leftLabel: '', rightLabel: '' }}>
    {Template.bind({})}
  </Story>
</Preview>

<Props of={Toggle} />
