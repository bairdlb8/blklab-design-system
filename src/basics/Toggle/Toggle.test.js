import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { axe, toHaveNoViolations } from 'jest-axe'
import Toggle from './Toggle'

expect.extend(toHaveNoViolations)

describe('<Toggle />', () => {
  describe('accessibility', () => {
    it('should not fail any accessibility tests', async () => {
      const { container } = render(<Toggle leftLabel="Test" rightLabel="Test" label="Testing" />)
      expect(await axe(container)).toHaveNoViolations()
    })
  })

  it('should take a snapshot', () => {
    const { asFragment } = render(<Toggle />)

    expect(asFragment(<Toggle />)).toMatchSnapshot()
  })
})
