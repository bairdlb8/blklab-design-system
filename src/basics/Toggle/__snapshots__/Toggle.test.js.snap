// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should take a snapshot 1`] = `
<DocumentFragment>
  <fieldset
    aria-label=""
  >
    <legend />
    <div
      class="ds-toggle"
      role="radiogroup"
    >
      <label
        class="sr-only"
        for="left"
      />
      <span
        class="ds-toggle__wrapper"
      >
        <input
          aria-label=""
          id="left"
          name=""
          type="radio"
        />
        <input
          aria-label=""
          id="right"
          name=""
          type="radio"
        />
        <span
          aria-hidden="true"
          class="ds-toggle__background"
        />
        <span
          aria-hidden="true"
          class="ds-toggle__switcher"
        />
      </span>
      <label
        class="sr-only"
        for="right"
      />
    </div>
  </fieldset>
</DocumentFragment>
`;
